File: Java/GameCharacter.java
/**
 * ゲームキャラクターを表す抽象クラスです。
 * @param initialLife ゲーム開始時点の体力
 * @param name ゲームキャラクターの名前
 */
abstract class GameCharacter {
  protected StandardIO standardIO = StandardIO.getInstance(); // 出力処理を担当するオブジェクトです。
  private final int MAX_LIFE = 5; // 全ゲームキャラクター共通となる体力の上限値です。
  private int life; // 体力です。
  private String name; // 名前です。
  protected final String[] ACTIONS = {"グー", "チョキ", "パー"}; // 選択できる行動の一覧です。
  protected String action; // 選択した行動です。
  GameCharacter(int initialLife, String name) {
    if (initialLife < 0 && initialLife > this.MAX_LIFE ? true : false) {
      System.out.println("[Error] ゲーム開始時点の初期値力値が不正です（" + initialLife + "）");
      System.exit(0);
    }
    this.life = initialLife;
    this.name = name;
  }

  /**
   * 行動を選択する抽象メソッドです。
   * 何らかの方法で行動を選択し、選択した行動を保持するフィールドに選択した行動を代入する処理を実装してください。
   */
  abstract void choiceAction();

  /**
   * ゲームの続行が可能か返答するメソッドです。
   * @return 続行可能ならtrue
   */
  boolean canGame() {
    return this.life > 0 ? true : false;
  }

  /**
   * 名前の横幅を返すメソッドです。
   * String.length()などで取得できる値とは異なり、半角文字を1、全角文字を2としてカウントした値を返します。
   */
  int getNameWidth() {
    char[] nameToCharArray = this.name.toCharArray();
    int nameWidth = 0;
    for (int i = 0; i < nameToCharArray.length; i += 1) {
      nameWidth += String.valueOf(nameToCharArray[i]).getBytes().length == 1 ? 1 : 2;
    }
    return nameWidth;
  }

  /**
   * 自身の状態をメッセージにして返すメソッドです。
   * メッセージは名前部分と体力部分から成ります。
   * 名前部分は相手の名前の長さと横幅を揃えるようにします。
   * 体力部分は記号を使って体力ゲージっぽく表現します。
   * @param opponentNameWidth 対戦相手の名前の横幅（半角文字を1、全角文字を2としてカウント）
   */
  void printStatusMessage(int opponentNameWidth) {
    StringBuilder namePart = new StringBuilder(this.name);
    for (int i = 0; i < opponentNameWidth - this.getNameWidth(); i += 1) {
      namePart.append(" ");
    }
    StringBuilder lifePart = new StringBuilder();
    for (int i = 0; i < this.life; i += 1) {
      lifePart.append("■");
    }
    for (int i = 0; i < this.MAX_LIFE - this.life; i += 1) {
      lifePart.append("□");
    }
    standardIO.println(namePart.toString() + ": " + lifePart.toString());
  }

  /**
   * 選択した行動を返すメソッドです。
   * @return 選択した行動
   */
  String getAction() {
    return this.action;
  }

  /**
   * ダメージ処理となるメソッドです。
   * 自身と相手の行動の組み合わせから、ダメージが発生するか確認し、ダメージが発生するならば体力を減らします。
   * @param opponentAction 対戦相手の行動
   */
  void damageProcess(String opponentAction) {
    if (this.action.equals(this.ACTIONS[0]) && opponentAction.equals(this.ACTIONS[2])
      || this.action.equals(this.ACTIONS[1]) && opponentAction.equals(this.ACTIONS[0])
      || this.action.equals(this.ACTIONS[2]) && opponentAction.equals(this.ACTIONS[1])
    ) {
      this.life -= 1;
      return;
    }
  }

  /**
   * 名前を返すメソッドです。
   * @return 名前
   */
  String getName() {
      return this.name;
  }
}

File: Kotlin/GameCharacter.kt
/**
 * ゲームキャラクターを表す抽象クラスです
 * 
 * @constructor
 * 
 * @param life プロパティlifeの初期値です
 * @param name プロパティnameの初期値です
 * 
 * @property maxLife 全ゲームキャラクター共通となる体力の上限値です
 * @property actionList ゲームキャラクターが選択できる行動のリストです
 * @property action 選択した行動です
 * @property life 体力です
 * @property name 名前です
 */
abstract class GameCharacter(private var life: Int, private val name: String) {
    private val maxLife = 5
    internal val actionList = listOf("グー", "チョキ", "パー")
    internal var action: String = ""    // Kotlinの文法上、メンバ変数は初期化が必要なので空文字を突っ込んでいます
    init {
        if (!(this.life in 0..this.maxLife)) throw Exception("[Error] ゲーム開始時点の体力が不正です（ $this.life ）")
    }

    /**
     * 行動を選択する抽象メソッドです
     */
    abstract fun choiceAction()

    /**
     * ゲームの続行が可能か返答するメソッドです
     * 
     * @return 続行可能ならtrueを返します
     */
    fun canGame(): Boolean = if (this.life > 0) true else false

    /**
     * 自身のステータスを表すメッセージを作成・出力するメソッドです
     * メッセージは名前部分と体力部分から成ります
     * → 名前部分は自身と相手の名前の「横幅」を揃えるようにします
     * → 体力部分は体力を記号を使って体力ゲージっぽく表現します
     * 
     * @param opponentNameWidth 対戦相手の名前の「横幅」です
     */
    fun printStatus(opponentNameWidth: Int) {
        var namePart = this.name
        repeat(opponentNameWidth - this.countNameWidth()) {
            namePart += " "
        }
        var lifePart = ""
        repeat(this.life) {
            lifePart += "■"
        }
        repeat(this.maxLife - this.life) {
            lifePart += "□"
        }
        StandardIO.println("$namePart : $lifePart")
    }

    /**
     * 自身の名前の「横幅」を返すメソッドです
     * ここでいう「横幅」とは半角文字を1、全角文字を2として数えた整数です
     * 
     * @return 名前の横幅を表す数値です
     */
    fun countNameWidth(): Int {
        var nameWidth = 0
        for (i in this.name.toCharArray()) nameWidth += if (i.toString().toByteArray().size == 1) 1 else 2
        return nameWidth
    }

    /**
     * 選択した行動を返すメソッドです
     * 
     * @return 選択した行動です
     */
    fun getAction(): String = this.action

    /**
     * ダメージ処理を実行するメソッドです
     * 自身と相手の行動を照らし合わせ、ダメージが発生するか確認します
     * ダメージが発生する状況であれば体力を減らします
     * 
     * @param opponentAction 対戦相手が選択した行動です
     */
    fun damageProcess(opponentAction: String) {
        if (!(this.actionList.contains(opponentAction))) throw Exception("[Error] 対戦相手の選択した行動が不正です（ $opponentAction ）")
        if ((this.action == this.actionList[0] && opponentAction == this.actionList[2])
            || (this.action == this.actionList[1] && opponentAction == this.actionList[0])
            || (this.action == this.actionList[2] && opponentAction == this.actionList[1])
        ) this.life -= 1
    }

    /**
     * 自身の名前を返すメソッドです
     * 
     * @return 自身の名前です
     */
    fun getName(): String = this.name
}